sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -511,6 +507,208 @@ class ConditionalHashJoinIterator(
+        } else {
+          updateTrackingMask(lazyRightMap)
+          Some(gatherer)
+        }
+      } finally {
+        maps.foreach(_.close())
+      }
+    }
+  }
+
+  override def close(): Unit = {
+    if (!closed) {
+      super.close()
+      compiledConditionRes.foreach(_.close())
+      rightTracker.foreach(_.close())
+    }
+  }
+
+  override def hasFinalBatch(): Option[ColumnarBatch] = {
+    rightTracker match {
+      case None => None
+      case Some(tracker) => {
+        val filteredBatch = withResource(tracker.releaseBatch()) { trackerBatch =>
+          withResource(GpuColumnVector.from(trackerBatch)) { trackerTab =>
+            val batch = built.getBatch
+            withResource(GpuColumnVector.from(batch)) { builtTable =>
+              withResource(builtTable.filter(trackerTab.getColumn(0))) { filterTab =>
+                GpuColumnVector.from(filterTab, GpuColumnVector.extractTypes(batch))
+              }
+            }
+          }
+        }
+        // Combine right columns with null columns for left side
+        val ret = withResource(filteredBatch) { rightBatch =>
+          val numFilterRows = rightBatch.numRows()
+          if (numFilterRows > 0) {
+            val leftColumns = streamAttributes.map { attr =>
+              GpuColumnVector.fromNull(numFilterRows, attr.dataType)
+            }
+            withResource(new ColumnarBatch(leftColumns.toArray, numFilterRows)) { leftBatch =>
+              Some(GpuColumnVector.combineColumns(leftBatch, rightBatch))
+            }
+          } else {
+            None
+          }
+        }
+        rightTracker = None
+        ret
+      }
+    }
+  }
+
+  private def trueColumnTable(numRows: Int): Table = {
+    withResource(Scalar.fromBool(true)) { trueScalar =>
+      withResource(ai.rapids.cudf.ColumnVector.fromScalar(trueScalar, numRows)) {
+        new Table(_)
+      }
+    }
+  }
+
+  // Create a boolean column to indicate which gather map rows are valid.
+  private def validIndexMask(gatherView: ColumnView): ColumnView = {
+    withResource(Scalar.fromInt(-numRightRows)) { minValue =>
+      withResource(Scalar.fromInt(numRightRows)) { maxValue =>
+        withResource(gatherView.greaterOrEqualTo(minValue)) { greaterEqualMin =>
+          withResource(gatherView.lessThan(maxValue)) { lessMax =>
+            lessMax.and(greaterEqualMin)
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Update the tracking mask.
+   */
+  private def updateTrackingMask(rightGatherMap: LazySpillableGatherMap): Unit = {
+    // Get the current tracking table, or all true table to start with
+    val rightTrackingTable = rightTracker.map { spillableBatch =>
+      withResource(spillableBatch.releaseBatch()) { trackingBatch =>
+        GpuColumnVector.from(trackingBatch)
+      }
+    }.getOrElse {
+      trueColumnTable(numRightRows)
+    }
+    // Filter the right gather map to remove invalid indices
+    val numGatherMapRows = rightGatherMap.getRowCount.toInt
```

Comment:
Are we guaranteed this doesn't truncate?  Unlike cudf columns, gather maps can be longer than 2^31 entries (which is why a gather map is not a column).  Unsure if we've already split the gather map to "fit" at this point yet or if this is the "raw" gather map that came out of a left outer join call in libcudf and has yet to be split if it's too large (e.g.: left outer join exploded).

Reply:
This is called via `createGatherer` from `SplittableJoinIterator.setupNextGatherersetupNextGatherer`.
It is called after the code that splits the stream side if we think the num rows will be too big.  But I was thinking this would be ok since this is the gather map for the build side, which should already be limited.  Is this an incorrect assumption?

Reply:
I didn't make a change for this.  Do you agree with my reasoning?   We are always using the build side gather map for this.

Reply:
yes we're fine here.